draft-xie-bidirectional-messaging-latest.md

```diff
@@ -30,63 +30,74 @@ author:
+available.
+
+To work around this limitation, several techniques are used, like long polling
+{{!RFC6202}}, WebSocket {{!RFC8441}}, and tunneling. These techniques generally
+layer application protocols on top of HTTP/2 and use HTTP/2 streams as transport
+connections. These solutions often defeat optimizations provided by HTTP/2.
+Multiplexing multiple parallel interactions onto one HTTP/2 stream reintroduces
+head of line blocking within that stream, application metadata is often
+encapsulated into DATA frames rather than HEADERS frames, so header compression
+is no longer effective, and adding an additional layer of framing to the user
+data reduces the wire efficiency advantages of HTTP/2 binary framing.
+
+For example, WebSocket over HTTP/2 frames user data both at the WebSocket layer
+and the HTTP/2 layer. This not only introduces additional overhead on the wire,
+but also complicates the data processing. These techniques also pose new
+operational challenges to intermediaries. If traffic from an entire user session
+is encapsulated into one HTTP/2 stream, this stream can last a very long time.
+Intermediaries may take long time to drain these streams. Unfortunately, GOAWAY
+only signals the remote endpoint to stop using the connection for new streams
+and often does not help drain the remaining open streams. Moreover,
+intermediaries have no visibility to user interactions, and lose the capability
+to collect telemetry metrics (e.g., time to the first/last byte of request and
+response) for services.
 
 In this draft, a new HTTP/2 frame is introduced which has the routing properties
-of PUSH_PROMISE frame and the bi-directionality of HEADERS frame. The extension
-provides several benefits: 1) after a HTTP/2 connection setup, a server can
-initiate streams to the client anytime, and the client can respond to the
-incoming streams accordingly. That is, the communication over HTTP/2 is
-bidirectional and symmetric. 2) All the HTTP/2 optimizations still apply.
-Intermediaries also have all the necessary meta data to accomplish their goals.
-3) Further, clients are able to group streams together for routing purposes,
+of a PUSH_PROMISE frame and the bi-directionality of a HEADERS frame. The
+extension provides several benefits:
+
+1. After a HTTP/2 connection is established, a
+server can initiate streams to the client at any time, and the client can respond to
+the incoming streams accordingly. That is, the communication over HTTP/2 is
+bidirectional and symmetric.
+
+2. All of the HTTP/2 technologies and optimizations still apply. Intermediaries
+also have all the necessary information to properly handle the communication
+between the client and the server.
+
+3. Clients are able to group streams together for routing purposes,
 such that each individual stream group can be used for a different service,
-within the same HTTP/2 connections.
+within the same HTTP/2 connection.
```

Comment:
It seems odd to me that we're mentioning groups for the first time as the last item in this list, when it seems to have a pretty big effect on what's going on. Shouldn't we be setting the stage earlier for what problem we solve with groups (for example, the intermediaries comment above could be resolved by something around this area)?

Reply:
Moving these to issues since we can tackle them later.

Reply:
Filed #6 to track this.